File: java/src/main/java/io/metaparticle/sync/Election.java
package io.metaparticle.sync;

public class Election implements LockListener, Runnable {
    private Lock lock;
    private Runnable electedAction;
    private Runnable terminateAction;
    // End signal is used to verify that the electedAction terminates properly when terminateAction
    // is called.
    private Object endSignal;
    private boolean done;
    private boolean running;
    
    public Election(String name, Runnable electedAction, Runnable terminateAction) {
        this(name, "http://localhost:13131", electedAction, terminateAction);
    }

    public Election(String name, String baseUrl, Runnable electedAction, Runnable terminateAction) {
        this.lock = new Lock(name, baseUrl);
        this.lock.setLockListener(this);
        this.electedAction = electedAction;
        this.terminateAction = terminateAction;
        this.endSignal = new Object();
    }

    public void shutdown() {
        running = false;
        synchronized (lock) {
            lock.notify();
        }
    }

    @Override
    public void run() {
        running = true;
        while (running) {
            try {
                lock.lock();
                synchronized (lock) {
                    lock.wait();
                }
            } catch (InterruptedException ex) {
                ex.printStackTrace();
            }
        }
        lock.unlock();
    }

    @Override
    public void lockAcquired() {
        new Thread(new Runnable() {
            public void run() {
                done = false;
                electedAction.run();
                synchronized (endSignal) {
                    endSignal.notify();
                    done = true;
                }
            }
        }).start();
    }

    @Override
    public void lockLost() {
        synchronized (endSignal) {
            terminateAction.run();
            // TODO: make this configurable?
            try {
                endSignal.wait(1000);
            } catch (InterruptedException ex) {
                ex.printStackTrace();
            }
            if (!done) {
                System.err.println("Master didn't terminate in expected time, force terminating.");
                System.exit(1);
            }
        }
        synchronized (lock) {
            lock.notify();
        }
    }

}

File: dotnet/Metaparticle.Sync/Election.cs
namespace Metaparticle.Sync {
    using System;
    using System.Threading;
    using System.Threading.Tasks;
    
    public class Election : LockListener {
        private Lock Lock;
        private Action ElectedAction;
        private Action TerminateAction;
        // End signal is used to verify that the electedAction terminates properly when terminateAction
        // is called.
        private Object EndSignal;
        private bool Done;
        private bool Running;
    
        public Election(string name, Action electedAction, Action terminateAction) :
            this(name, "http://localhost:13131", electedAction, terminateAction) {}

        public Election(string name, string baseUrl, Action electedAction, Action terminateAction) {
            this.Lock = new Lock(name, baseUrl);
            this.Lock.Listener = this;
            this.ElectedAction = electedAction;
            this.TerminateAction = terminateAction;
            this.EndSignal = new Object();
        }

        public void Shutdown() {
            Running = false;
            lock (Lock) {
                Monitor.Pulse(Lock);
            }
        }

        public void Run() {
            Running = true;
            while (Running) {
                Lock.WaitOne();
                lock (Lock) {
                    Monitor.Wait(Lock);
                }
            }
            Lock.Release();
        }

        public void LockAcquired() {
            Task.Run(() => {
                Done = false;
                ElectedAction();
                lock (EndSignal) {
                    Monitor.Pulse(EndSignal);
                    Done = true;
                }
            });
        }

        public void LockLost() {
            lock (EndSignal) {
                TerminateAction();
                // TODO: make this configurable?
                Task.Delay(1000).Wait();
                if (!Done) {
                    Console.WriteLine("Master didn't terminate in expected time, force terminating.");
                    //Application.exit(1);
                }
            }
            lock (Lock) {
                Monitor.Pulse(Lock);
            }
        }
    }
}